From cea067134444d228979fee30067ecde9acf9db0e Mon Sep 17 00:00:00 2001
From: Giuseppe Cavallaro <peppe.cavallaro@st.com>
Date: Wed, 15 Jul 2009 11:22:55 +0200
Subject: [PATCH] sh: add assembly memchr from Kernel

Faster assembly implementation of memchr taken from Linux kernel.
Adapted to Glibc.

Signed-off-by: Giuseppe Cavallaro <peppe.cavallaro@st.com>
---
 sysdeps/sh/memchr.S |   39 +++++++++++++++++++++++++++++++++++++++
 1 files changed, 39 insertions(+), 0 deletions(-)
 create mode 100644 sysdeps/sh/memchr.S

diff --git a/sysdeps/sh/memchr.S b/sysdeps/sh/memchr.S
new file mode 100644
index 0000000..abc2dfa
--- /dev/null
+++ b/sysdeps/sh/memchr.S
@@ -0,0 +1,39 @@
+/* Copyright (C) 1999  Niibe Yutaka
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, write to the Free
+   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+   02111-1307 USA.  */
+
+/*
+ * void *memchr(const void *s, int c, size_t n);
+ */
+
+#include <sysdep.h>
+
+ENTRY(memchr)
+	tst	r6,r6
+	bt/s	2f
+	 exts.b	r5,r5
+1:	mov.b	@r4,r1
+	cmp/eq	r1,r5
+	bt/s	3f
+	 dt	r6
+	bf/s	1b
+	 add	#1,r4
+2:	mov	#0,r4
+3:	rts
+	 mov	r4,r0
+END(memchr)
+libc_hidden_builtin_def (memchr)
-- 
1.6.0.6

